File: packages/grua_core/lib/src/domain/enums.dart
/// Domain enumerations shared by the three apps and mirrored in
/// `functions/src/lib/enums.ts`.
///
/// Every enum here follows the same contract: a `wire` string that is the only
/// thing ever written to Firestore, and a tolerant `fromWire` that resolves an
/// unrecognised value to an `unknown` member instead of throwing. That matters
/// because an app on a customer's phone can be months behind the backend, and a
/// chofer whose app crashes on a new status is a chofer who cannot work.
library;

import 'package:json_annotation/json_annotation.dart';

/// Resolves [wire] against [values], falling back to [fallback].
T _resolve<T>(List<T> values, String? wire, String Function(T) key, T fallback) {
  if (wire == null) return fallback;
  for (final value in values) {
    if (key(value) == wire) return value;
  }
  return fallback;
}

// ---------------------------------------------------------------------------
// Identity
// ---------------------------------------------------------------------------

enum UserRole {
  @JsonValue('client')
  client('client'),
  @JsonValue('driver')
  driver('driver'),
  @JsonValue('admin')
  admin('admin'),
  @JsonValue('ops')
  ops('ops'),
  @JsonValue('unknown')
  unknown('unknown');

  const UserRole(this.wire);

  final String wire;

  static UserRole fromWire(String? wire) =>
      _resolve(UserRole.values, wire, (v) => v.wire, UserRole.unknown);

  bool get isStaff => this == UserRole.admin || this == UserRole.ops;
}

/// Lifecycle of a chofer's account, controlled entirely from the admin panel.
enum DriverStatus {
  /// Created but not yet cleared to work — usually missing documents.
  @JsonValue('inactive')
  inactive('inactive'),

  /// Cleared to go online and receive offers.
  @JsonValue('active')
  active('active'),

  /// Blocked by an admin. Cannot log in to work.
  @JsonValue('suspended')
  suspended('suspended'),

  @JsonValue('unknown')
  unknown('unknown');

  const DriverStatus(this.wire);

  final String wire;

  static DriverStatus fromWire(String? wire) =>
      _resolve(DriverStatus.values, wire, (v) => v.wire, DriverStatus.unknown);

  bool get canWork => this == DriverStatus.active;
}

/// What the chofer is doing right now, as published to `/live/{driverId}`.
enum DriverLiveState {
  /// Online and dispatchable.
  @JsonValue('idle')
  idle('idle'),

  /// Online but already committed to a service.
  @JsonValue('on_service')
  onService('on_service'),

  @JsonValue('unknown')
  unknown('unknown');

  const DriverLiveState(this.wire);

  final String wire;

  static DriverLiveState fromWire(String? wire) =>
      _resolve(DriverLiveState.values, wire, (v) => v.wire, DriverLiveState.unknown);
}

// ---------------------------------------------------------------------------
// Fleet
// ---------------------------------------------------------------------------

/// The kind of grúa required. This is the field dispatch filters on, so an
/// unknown value must never be dispatchable.
enum TruckType {
  /// Flatbed. Required for anything that cannot roll or must not be towed on
  /// its own wheels.
  @JsonValue('plataforma')
  plataforma('plataforma', 'Plataforma'),

  /// Hook and chain / wheel-lift. The everyday tow.
  @JsonValue('gancho')
  gancho('gancho', 'Gancho'),

  /// Heavy recovery for trucks and buses.
  @JsonValue('pesada')
  pesada('pesada', 'Grúa pesada'),

  @JsonValue('unknown')
  unknown('unknown', 'Desconocido');

  const TruckType(this.wire, this.label);

  final String wire;

  /// Spanish label shown to users in the Dominican Republic.
  final String label;

  static TruckType fromWire(String? wire) =>
      _resolve(TruckType.values, wire, (v) => v.wire, TruckType.unknown);

  bool get isDispatchable => this != TruckType.unknown;
}

/// The customer's vehicle class, used to infer [TruckType].
enum VehicleType {
  @JsonValue('sedan')
  sedan('sedan', 'Carro / Sedán'),
  @JsonValue('suv')
  suv('suv', 'Jeepeta / SUV'),
  @JsonValue('camioneta')
  camioneta('camioneta', 'Camioneta'),
  @JsonValue('camion')
  camion('camion', 'Camión / Autobús'),
  @JsonValue('motor')
  motor('motor', 'Motor'),
  @JsonValue('unknown')
  unknown('unknown', 'Otro');

  const VehicleType(this.wire, this.label);

  final String wire;
  final String label;

  static VehicleType fromWire(String? wire) =>
      _resolve(VehicleType.values, wire, (v) => v.wire, VehicleType.unknown);
}

/// Why the vehicle needs a grúa. Drives both pricing and truck-type inference.
enum VehicleCondition {
  @JsonValue('no_arranca')
  noArranca('no_arranca', 'No arranca'),
  @JsonValue('accidentado')
  accidentado('accidentado', 'Accidentado'),
  @JsonValue('ruedas_bloqueadas')
  ruedasBloqueadas('ruedas_bloqueadas', 'Ruedas bloqueadas'),
  @JsonValue('volcado')
  volcado('volcado', 'Volcado'),
  @JsonValue('sin_combustible')
  sinCombustible('sin_combustible', 'Sin combustible'),
  @JsonValue('goma_pinchada')
  gomaPinchada('goma_pinchada', 'Goma pinchada'),
  @JsonValue('unknown')
  unknown('unknown', 'Otro problema');

  const VehicleCondition(this.wire, this.label);

  final String wire;
  final String label;

  static VehicleCondition fromWire(String? wire) =>
      _resolve(VehicleCondition.values, wire, (v) => v.wire, VehicleCondition.unknown);

  /// A vehicle that cannot roll on its own wheels needs a flatbed.
  bool get requiresFlatbed =>
      this == VehicleCondition.volcado ||
      this == VehicleCondition.accidentado ||
      this == VehicleCondition.ruedasBloqueadas;
}

/// Documents a chofer must keep current to stay `active`.
enum DriverDocumentType {
  @JsonValue('licencia')
  licencia('licencia', 'Licencia de conducir', required: true),
  @JsonValue('cedula')
  cedula('cedula', 'Cédula', required: true),
  @JsonValue('seguro')
  seguro('seguro', 'Seguro del vehículo', required: true),
  @JsonValue('marbete')
  marbete('marbete', 'Marbete', required: true),
  @JsonValue('matricula')
  matricula('matricula', 'Matrícula', required: true),
  @JsonValue('certificado_medico')
  certificadoMedico('certificado_medico', 'Certificado médico', required: false),
  @JsonValue('unknown')
  unknown('unknown', 'Documento', required: false);

  const DriverDocumentType(this.wire, this.label, {required this.required});

  final String wire;
  final String label;

  /// When a required document expires, the chofer is forced offline.
  final bool required;

  static DriverDocumentType fromWire(String? wire) =>
      _resolve(DriverDocumentType.values, wire, (v) => v.wire, DriverDocumentType.unknown);
}

enum DocumentReviewState {
  @JsonValue('pending')
  pending('pending', 'Pendiente'),
  @JsonValue('verified')
  verified('verified', 'Verificado'),
  @JsonValue('rejected')
  rejected('rejected', 'Rechazado'),
  @JsonValue('expired')
  expired('expired', 'Vencido'),
  @JsonValue('unknown')
  unknown('unknown', 'Desconocido');

  const DocumentReviewState(this.wire, this.label);

  final String wire;
  final String label;

  static DocumentReviewState fromWire(String? wire) =>
      _resolve(DocumentReviewState.values, wire, (v) => v.wire, DocumentReviewState.unknown);
}

// ---------------------------------------------------------------------------
// Service lifecycle
// ---------------------------------------------------------------------------

/// The eleven service states. Transitions are enforced server-side; this enum
/// exists so the apps can render the right screen, never to decide a change.
enum ServiceStatus {
  /// Created and looking for a chofer.
  @JsonValue('pending_dispatch')
  pendingDispatch('pending_dispatch', 'Buscando grúa'),

  /// One chofer is holding an exclusive 25-second offer.
  @JsonValue('offered')
  offered('offered', 'Buscando grúa'),

  /// A chofer took it and is on the way.
  @JsonValue('accepted')
  accepted('accepted', 'Grúa en camino'),

  /// The chofer pressed "Llegué".
  @JsonValue('arrived')
  arrived('arrived', 'Tu grúa llegó'),

  /// Vehicle loaded, heading to the destination.
  @JsonValue('in_progress')
  inProgress('in_progress', 'En camino al destino'),

  /// The chofer pressed "Finalizar". Payment may still be settling.
  @JsonValue('completed')
  completed('completed', 'Servicio completado'),

  /// Paid and invoiced. Terminal.
  @JsonValue('closed')
  closed('closed', 'Servicio cerrado'),

  /// The cascade gave up; a dispatcher must assign by hand.
  @JsonValue('needs_manual')
  needsManual('needs_manual', 'Asignando grúa'),

  /// Cancelled by client, chofer, admin or the system. Terminal.
  @JsonValue('cancelled')
  cancelled('cancelled', 'Servicio cancelado'),

  /// Nobody was ever assigned within the dispatch window. Terminal.
  @JsonValue('expired')
  expired('expired', 'Servicio expirado'),

  /// Something went wrong that needs a human. Terminal.
  @JsonValue('failed')
  failed('failed', 'Servicio con problema'),

  @JsonValue('unknown')
  unknown('unknown', 'Estado desconocido');

  const ServiceStatus(this.wire, this.label);

  final String wire;

  /// Customer-facing es-DO label. Note that `offered` deliberately reads the
  /// same as `pending_dispatch`: the client should not see the cascade churn.
  final String label;

  static ServiceStatus fromWire(String? wire) =>
      _resolve(ServiceStatus.values, wire, (v) => v.wire, ServiceStatus.unknown);

  static const Set<ServiceStatus> terminal = {
    ServiceStatus.closed,
    ServiceStatus.cancelled,
    ServiceStatus.expired,
    ServiceStatus.failed,
  };

  /// States in which a client is considered to have a service in flight and
  /// may not request another.
  static const Set<ServiceStatus> active = {
    ServiceStatus.pendingDispatch,
    ServiceStatus.offered,
    ServiceStatus.accepted,
    ServiceStatus.arrived,
    ServiceStatus.inProgress,
    ServiceStatus.completed,
    ServiceStatus.needsManual,
  };

  /// States where chat and calling between the two parties are open.
  static const Set<ServiceStatus> contactOpen = {
    ServiceStatus.accepted,
    ServiceStatus.arrived,
    ServiceStatus.inProgress,
  };

  bool get isTerminal => terminal.contains(this);

  bool get isActive => active.contains(this);

  /// The office's name for the state. [label] is written for the customer —
  /// "Tu grúa llegó", and `offered` hidden behind "Buscando grúa" — which is
  /// the wrong voice for a dispatcher reading a list of every job.
  String get officeLabel => switch (this) {
        ServiceStatus.pendingDispatch => 'Buscando chofer',
        ServiceStatus.offered => 'Ofrecido a chofer',
        ServiceStatus.accepted => 'Chofer en camino',
        ServiceStatus.arrived => 'Chofer en el punto',
        ServiceStatus.inProgress => 'Remolcando',
        ServiceStatus.completed => 'Completado',
        ServiceStatus.closed => 'Cerrado',
        ServiceStatus.needsManual => 'Requiere asignación',
        ServiceStatus.cancelled => 'Cancelado',
        ServiceStatus.expired => 'Expirado',
        ServiceStatus.failed => 'Con problema',
        ServiceStatus.unknown => 'Desconocido',
      };

  /// True once a specific chofer owns the job.
  bool get hasDriver => const {
        ServiceStatus.accepted,
        ServiceStatus.arrived,
        ServiceStatus.inProgress,
        ServiceStatus.completed,
        ServiceStatus.closed,
      }.contains(this);

  /// The client may cancel right up until the vehicle is loaded.
  bool get isCancellableByClient => const {
        ServiceStatus.pendingDispatch,
        ServiceStatus.offered,
        ServiceStatus.needsManual,
        ServiceStatus.accepted,
        ServiceStatus.arrived,
      }.contains(this);

  bool get allowsContact => contactOpen.contains(this);
}

/// Names of the server callables that move a service between states. Keeping
/// them here means the apps and the tests cannot drift from the function names.
enum ServiceEventName {
  @JsonValue('requestService')
  requestService('requestService'),
  @JsonValue('dispatchNext')
  dispatchNext('dispatchNext'),
  @JsonValue('acceptService')
  acceptService('acceptService'),
  @JsonValue('rejectService')
  rejectService('rejectService'),
  @JsonValue('expireOffer')
  expireOffer('expireOffer'),
  @JsonValue('noDriversFound')
  noDriversFound('noDriversFound'),
  @JsonValue('assignServiceManually')
  assignServiceManually('assignServiceManually'),
  @JsonValue('markArrived')
  markArrived('markArrived'),
  @JsonValue('startService')
  startService('startService'),
  @JsonValue('completeService')
  completeService('completeService'),
  @JsonValue('confirmCashCollected')
  confirmCashCollected('confirmCashCollected'),
  @JsonValue('closeService')
  closeService('closeService'),
  @JsonValue('cancelService')
  cancelService('cancelService'),
  @JsonValue('cancelByDriver')
  cancelByDriver('cancelByDriver'),
  @JsonValue('failService')
  failService('failService'),
  @JsonValue('unknown')
  unknown('unknown');

  const ServiceEventName(this.wire);

  final String wire;

  static ServiceEventName fromWire(String? wire) =>
      _resolve(ServiceEventName.values, wire, (v) => v.wire, ServiceEventName.unknown);
}

/// State of a single dispatch offer to one chofer.
enum OfferState {
  @JsonValue('sent')
  sent('sent'),
  @JsonValue('accepted')
  accepted('accepted'),
  @JsonValue('rejected')
  rejected('rejected'),
  @JsonValue('expired')
  expired('expired'),
  @JsonValue('cancelled')
  cancelled('cancelled'),
  @JsonValue('unknown')
  unknown('unknown');

  const OfferState(this.wire);

  final String wire;

  static OfferState fromWire(String? wire) =>
      _resolve(OfferState.values, wire, (v) => v.wire, OfferState.unknown);

  bool get isOpen => this == OfferState.sent;
}

enum AssignmentMode {
  @JsonValue('auto')
  auto('auto', 'Automático'),
  @JsonValue('manual')
  manual('manual', 'Manual'),
  @JsonValue('unknown')
  unknown('unknown', 'Desconocido');

  const AssignmentMode(this.wire, this.label);

  final String wire;
  final String label;

  static AssignmentMode fromWire(String? wire) =>
      _resolve(AssignmentMode.values, wire, (v) => v.wire, AssignmentMode.unknown);
}

enum CancelledBy {
  @JsonValue('client')
  client('client', 'Cliente'),
  @JsonValue('driver')
  driver('driver', 'Chofer'),
  @JsonValue('admin')
  admin('admin', 'Administración'),
  @JsonValue('system')
  system('system', 'Sistema'),
  @JsonValue('unknown')
  unknown('unknown', 'Desconocido');

  const CancelledBy(this.wire, this.label);

  final String wire;
  final String label;

  static CancelledBy fromWire(String? wire) =>
      _resolve(CancelledBy.values, wire, (v) => v.wire, CancelledBy.unknown);
}

/// Fixed reasons a chofer may give for dropping a job. Free text is not
/// accepted because these feed the admin's abuse flags.
enum DriverCancelReason {
  @JsonValue('vehicle_breakdown')
  vehicleBreakdown('vehicle_breakdown', 'Avería de la grúa'),
  @JsonValue('wrong_truck_type')
  wrongTruckType('wrong_truck_type', 'Tipo de grúa incorrecto'),
  @JsonValue('client_not_present')
  clientNotPresent('client_not_present', 'El cliente no está en el lugar'),
  @JsonValue('client_refused')
  clientRefused('client_refused', 'El cliente rechazó el servicio'),
  @JsonValue('inaccessible_location')
  inaccessibleLocation('inaccessible_location', 'No puedo llegar al lugar'),
  @JsonValue('unsafe_location')
  unsafeLocation('unsafe_location', 'Lugar inseguro'),
  @JsonValue('emergency')
  emergency('emergency', 'Emergencia personal'),
  @JsonValue('other')
  other('other', 'Otro motivo'),
  @JsonValue('unknown')
  unknown('unknown', 'Desconocido');

  const DriverCancelReason(this.wire, this.label);

  final String wire;
  final String label;

  static DriverCancelReason fromWire(String? wire) =>
      _resolve(DriverCancelReason.values, wire, (v) => v.wire, DriverCancelReason.unknown);
}

// ---------------------------------------------------------------------------
// Money
// ---------------------------------------------------------------------------

enum PaymentMethod {
  @JsonValue('card')
  card('card', 'Tarjeta'),
  @JsonValue('cash')
  cash('cash', 'Efectivo'),
  @JsonValue('unknown')
  unknown('unknown', 'Desconocido');

  const PaymentMethod(this.wire, this.label);

  final String wire;
  final String label;

  static PaymentMethod fromWire(String? wire) =>
      _resolve(PaymentMethod.values, wire, (v) => v.wire, PaymentMethod.unknown);
}

enum PaymentStatus {
  /// Cash job, or a card job before the hold is placed.
  @JsonValue('none')
  none('none', 'Sin procesar'),

  /// Card hold placed at accept time.
  @JsonValue('authorized')
  authorized('authorized', 'Autorizado'),

  /// Hold captured at completion.
  @JsonValue('captured')
  captured('captured', 'Cobrado'),

  /// Authorization or capture was declined.
  @JsonValue('failed')
  failed('failed', 'Rechazado'),

  @JsonValue('refunded')
  refunded('refunded', 'Reembolsado'),

  /// Completed cash job, chofer has not confirmed collection yet.
  @JsonValue('cash_pending')
  cashPending('cash_pending', 'Cobro en efectivo pendiente'),

  @JsonValue('cash_collected')
  cashCollected('cash_collected', 'Efectivo recibido'),

  @JsonValue('unknown')
  unknown('unknown', 'Desconocido');

  const PaymentStatus(this.wire, this.label);

  final String wire;
  final String label;

  static PaymentStatus fromWire(String? wire) =>
      _resolve(PaymentStatus.values, wire, (v) => v.wire, PaymentStatus.unknown);

  bool get isSettled =>
      this == PaymentStatus.captured || this == PaymentStatus.cashCollected;

  /// A card service cannot start until the hold is in place.
  bool get blocksServiceStart => this == PaymentStatus.failed || this == PaymentStatus.none;
}

/// Dominican tax receipt types (Números de Comprobante Fiscal).
enum NcfType {
  /// Crédito fiscal — for a customer with an RNC who will deduct the ITBIS.
  @JsonValue('01')
  creditoFiscal('01', 'Crédito fiscal'),

  /// Consumo — the default for individuals.
  @JsonValue('02')
  consumo('02', 'Consumo'),

  @JsonValue('unknown')
  unknown('unknown', 'Desconocido');

  const NcfType(this.wire, this.label);

  final String wire;
  final String label;

  static NcfType fromWire(String? wire) =>
      _resolve(NcfType.values, wire, (v) => v.wire, NcfType.unknown);
}

File: functions/src/lib/enums.ts
/**
 * Wire values, mirrored from `packages/grua_core/lib/src/domain/enums.dart`.
 *
 * These are the strings actually stored in Firestore, so the two files must
 * agree exactly. They are duplicated rather than generated because a codegen
 * step between Dart and TypeScript would be a build dependency for the sake of
 * ~60 string literals — but a change on either side without the other is a
 * silent data bug, so `enums.test.ts` asserts the sets match the Dart source.
 */

export const ServiceStatus = {
  pendingDispatch: 'pending_dispatch',
  offered: 'offered',
  accepted: 'accepted',
  arrived: 'arrived',
  inProgress: 'in_progress',
  completed: 'completed',
  closed: 'closed',
  needsManual: 'needs_manual',
  cancelled: 'cancelled',
  expired: 'expired',
  failed: 'failed',
} as const;

export type ServiceStatus = (typeof ServiceStatus)[keyof typeof ServiceStatus];

/** States in which a customer is considered to have a tow in flight. */
export const ACTIVE_STATUSES: readonly ServiceStatus[] = [
  ServiceStatus.pendingDispatch,
  ServiceStatus.offered,
  ServiceStatus.accepted,
  ServiceStatus.arrived,
  ServiceStatus.inProgress,
  ServiceStatus.completed,
  ServiceStatus.needsManual,
];

export const TERMINAL_STATUSES: readonly ServiceStatus[] = [
  ServiceStatus.closed,
  ServiceStatus.cancelled,
  ServiceStatus.expired,
  ServiceStatus.failed,
];

/** States where chat and calling between the two parties are open. */
export const CONTACT_OPEN_STATUSES: readonly ServiceStatus[] = [
  ServiceStatus.accepted,
  ServiceStatus.arrived,
  ServiceStatus.inProgress,
];

export const ServiceEventName = {
  requestService: 'requestService',
  dispatchNext: 'dispatchNext',
  acceptService: 'acceptService',
  rejectService: 'rejectService',
  expireOffer: 'expireOffer',
  noDriversFound: 'noDriversFound',
  assignServiceManually: 'assignServiceManually',
  confirmHeavyService: 'confirmHeavyService',
  markArrived: 'markArrived',
  startService: 'startService',
  completeService: 'completeService',
  confirmCashCollected: 'confirmCashCollected',
  closeService: 'closeService',
  cancelService: 'cancelService',
  cancelByDriver: 'cancelByDriver',
  failService: 'failService',

  // Money, logged beside the transitions. None of these move the status.
  choosePaymentMethod: 'choosePaymentMethod',
  paymentAuthorized: 'paymentAuthorized',
  paymentCaptured: 'paymentCaptured',
  paymentFailed: 'paymentFailed',
  paymentVoided: 'paymentVoided',
} as const;

export type ServiceEventName =
  (typeof ServiceEventName)[keyof typeof ServiceEventName];

export const UserRole = {
  client: 'client',
  driver: 'driver',
  admin: 'admin',
  ops: 'ops',
  /** A person working for an insurance company that is billed monthly. */
  insurer: 'insurer',
} as const;

export type UserRole = (typeof UserRole)[keyof typeof UserRole];

/**
 * What a person can do inside their insurance company.
 *
 * Both create and follow tows. Only a manager adds or removes the company's
 * people, and only a manager sees its invoices.
 */
export const InsurerRole = {
  manager: 'manager',
  operator: 'operator',
} as const;

export type InsurerRole = (typeof InsurerRole)[keyof typeof InsurerRole];

/** Whether an insurance company may use the platform at all. */
export const InsurerStatus = {
  active: 'active',
  suspended: 'suspended',
} as const;

export type InsurerStatus = (typeof InsurerStatus)[keyof typeof InsurerStatus];

export const DriverStatus = {
  inactive: 'inactive',
  active: 'active',
  suspended: 'suspended',
} as const;

export type DriverStatus = (typeof DriverStatus)[keyof typeof DriverStatus];

/** Paperwork a chofer must keep current. Mirrors `DriverDocumentType` in Dart. */
export const DriverDocumentType = {
  licencia: 'licencia',
  licenciaReverso: 'licencia_reverso',
  cedula: 'cedula',
  seguro: 'seguro',
  marbete: 'marbete',
  matricula: 'matricula',
  certificadoMedico: 'certificado_medico',
} as const;

export type DriverDocumentType =
  (typeof DriverDocumentType)[keyof typeof DriverDocumentType];

export const DocumentReviewState = {
  pending: 'pending',
  verified: 'verified',
  rejected: 'rejected',
  expired: 'expired',
} as const;

export type DocumentReviewState =
  (typeof DocumentReviewState)[keyof typeof DocumentReviewState];

/**
 * Where a self-registered chofer's licence check stands. Mirrors
 * `LicenseVerificationState` in Dart. Choferes the office opened never get
 * one: the office saw their papers in person.
 */
export const LicenseVerificationState = {
  /** Registered; the two photos have not been submitted for checking yet. */
  awaitingDocuments: 'awaiting_documents',
  processing: 'processing',
  /** Passed every check. The account still waits for an admin to activate it. */
  verified: 'verified',
  /** Failed a check the chofer can fix with new photos. */
  rejected: 'rejected',
  /** Needs a person: an unclear result, a suspected edit, or too many tries. */
  manualReview: 'manual_review',
} as const;

export type LicenseVerificationState =
  (typeof LicenseVerificationState)[keyof typeof LicenseVerificationState];

export const DriverLiveState = {
  idle: 'idle',
  onService: 'on_service',
} as const;

export type DriverLiveState =
  (typeof DriverLiveState)[keyof typeof DriverLiveState];

export const TruckType = {
  plataforma: 'plataforma',
  gancho: 'gancho',
  pesada: 'pesada',
} as const;

export type TruckType = (typeof TruckType)[keyof typeof TruckType];

export const VehicleType = {
  /** Carro. */
  sedan: 'sedan',
  /** Jeepeta. */
  suv: 'suv',
  camioneta: 'camioneta',
  /** Camión de 2 ejes. */
  camion: 'camion',
  /** Patana / tráiler. */
  patana: 'patana',
  equipoPesado: 'equipo_pesado',
  motor: 'motor',
} as const;

export type VehicleType = (typeof VehicleType)[keyof typeof VehicleType];

/**
 * Vehículos pesados: a special grúa, a price that is only an estimate, and an
 * operator who confirms both before anybody drives out.
 */
export const HEAVY_VEHICLE_TYPES: readonly VehicleType[] = [
  VehicleType.camion,
  VehicleType.patana,
  VehicleType.equipoPesado,
];

export const isHeavyVehicle = (type: VehicleType | string | undefined): boolean =>
  HEAVY_VEHICLE_TYPES.includes(type as VehicleType);

/** Where a heavy request stands with the operator. */
export const OperatorReviewState = {
  pending: 'pending',
  confirmed: 'confirmed',
} as const;

export type OperatorReviewState =
  (typeof OperatorReviewState)[keyof typeof OperatorReviewState];

export const VehicleCondition = {
  noArranca: 'no_arranca',
  accidentado: 'accidentado',
  ruedasBloqueadas: 'ruedas_bloqueadas',
  volcado: 'volcado',
  sinCombustible: 'sin_combustible',
  gomaPinchada: 'goma_pinchada',
} as const;

export type VehicleCondition =
  (typeof VehicleCondition)[keyof typeof VehicleCondition];

/** Conditions where the vehicle cannot roll on its own wheels. */
export const FLATBED_CONDITIONS: readonly VehicleCondition[] = [
  VehicleCondition.volcado,
  VehicleCondition.accidentado,
  VehicleCondition.ruedasBloqueadas,
];

export const OfferState = {
  sent: 'sent',
  accepted: 'accepted',
  rejected: 'rejected',
  expired: 'expired',
  cancelled: 'cancelled',
} as const;

export type OfferState = (typeof OfferState)[keyof typeof OfferState];

/**
 * How the customer pays. `pending` until they choose, which they do when the
 * chofer arrives: the price can still change on the way (an operator's
 * confirmation, a longer wait), and a card is only held once there is a truck
 * at the curb.
 */
export const PaymentMethod = {
  card: 'card',
  cash: 'cash',
  pending: 'pending',
  /** Nobody pays at the roadside: the insurance company is billed monthly. */
  insurer: 'insurer',
} as const;
export type PaymentMethod = (typeof PaymentMethod)[keyof typeof PaymentMethod];

export const PaymentStatus = {
  none: 'none',
  /** The card is held for the job; nothing charged yet. */
  authorized: 'authorized',
  /** Charged. Shown to everyone as "Pagado". */
  captured: 'captured',
  failed: 'failed',
  refunded: 'refunded',
  /** A hold released without charging: cancelled, or switched to cash. */
  voided: 'voided',
  cashPending: 'cash_pending',
  cashCollected: 'cash_collected',
  /** An insurer's tow, done, waiting for the month's invoice. */
  toInvoice: 'to_invoice',
  /** An insurer's tow on a monthly invoice, whose id is in `invoiceId`. */
  invoiced: 'invoiced',
} as const;

export type PaymentStatus = (typeof PaymentStatus)[keyof typeof PaymentStatus];

export const AssignmentMode = { auto: 'auto', manual: 'manual' } as const;
export type AssignmentMode =
  (typeof AssignmentMode)[keyof typeof AssignmentMode];

export const CancelledBy = {
  client: 'client',
  /** A person of the insurance company that ordered the tow. */
  insurer: 'insurer',
  driver: 'driver',
  admin: 'admin',
  system: 'system',
} as const;

export type CancelledBy = (typeof CancelledBy)[keyof typeof CancelledBy];

export const DriverCancelReason = {
  vehicleBreakdown: 'vehicle_breakdown',
  wrongTruckType: 'wrong_truck_type',
  clientNotPresent: 'client_not_present',
  clientRefused: 'client_refused',
  inaccessibleLocation: 'inaccessible_location',
  unsafeLocation: 'unsafe_location',
  emergency: 'emergency',
  other: 'other',
} as const;

export type DriverCancelReason =
  (typeof DriverCancelReason)[keyof typeof DriverCancelReason];

export const NcfType = { creditoFiscal: '01', consumo: '02' } as const;
export type NcfType = (typeof NcfType)[keyof typeof NcfType];

/**
 * The truck type a vehicle needs.
 *
 * Ported from `ServiceVehicle.inferredTruckType`. The app runs the same rule to
 * show a price before the customer commits, but this is the authority — a
 * modified client claiming a `gancho` tow for a rolled-over car would otherwise
 * get the cheap rate and a truck that cannot lift it.
 */
export function inferTruckType(
  vehicleType: VehicleType,
  condition: VehicleCondition,
): TruckType {
  // First, whatever the condition: a flatbed cannot lift a camión, let alone a
  // patana or a loader, rolled over or not.
  if (isHeavyVehicle(vehicleType)) return TruckType.pesada;
  if (FLATBED_CONDITIONS.includes(condition)) return TruckType.plataforma;
  return TruckType.gancho;
}

/**
 * Which trucks can actually do a job that asks for `required`, best first.
 *
 * Dispatch used to demand an exact match, and that is not how a yard works. A
 * plataforma carries the whole vehicle, so it can do anything a gancho can —
 * refusing it meant a customer with a car that would not start watched
 * "Buscando grúa" for six minutes while an idle flatbed sat two streets away,
 * and the job ended up on a dispatcher's desk as `needs_manual`.
 *
 * Not the other way round: a gancho tows on the vehicle's own wheels, which is
 * exactly what a flipped or wheel-locked car cannot do. And a pesada is for
 * trucks and buses; nothing substitutes for it and it substitutes for nothing,
 * because sending a heavy wrecker to a sedan is the wrong truck at the wrong
 * price.
 */
export function trucksThatCanServe(required: TruckType): readonly TruckType[] {
  switch (required) {
    case TruckType.gancho:
      return [TruckType.gancho, TruckType.plataforma];
    case TruckType.plataforma:
      return [TruckType.plataforma];
    default:
      return [required];
  }
}
